native/packages/react-native-bpk-component-carousel/src/BpkCarousel.js

```diff
@@ -0,0 +1,116 @@
+/*
+ * Backpack - Skyscanner's Design System
+ *
+ * Copyright 2018 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* @flow */
+
+import React, { Fragment, type Element, type ChildrenArray } from 'react';
+import PropTypes from 'prop-types';
+import { ScrollView, type StyleObj, ViewPropTypes } from 'react-native';
+import typeof BpkCarouselItem from './BpkCarouselItem';
+
+export type Props = {
+  accessibilityLabel: string | ((number, number) => string),
+  children: ChildrenArray<Element<BpkCarouselItem>>,
+  style: ?StyleObj,
+  renderPagination: ?(number, number) => Element<any>,
+};
+
+type State = {
+  currentIndex: number,
+  width: ?number,
+  height: ?number,
+};
+
+class BpkCarousel extends React.Component<Props, State> {
+  static propTypes = {
+    accessibilityLabel: PropTypes.oneOfType([PropTypes.string, PropTypes.func])
+      .isRequired,
+    children: PropTypes.node.isRequired,
+    style: ViewPropTypes.style,
+    renderPagination: PropTypes.func,
+  };
+
+  static defaultProps = {
+    style: null,
+    renderPagination: null,
+  };
+
+  constructor(props: Props) {
+    super(props);
+    this.state = { width: null, height: null, currentIndex: 0 };
+  }
+
+  onLayout = (event: any) => {
+    this.setState({
+      width: event.nativeEvent.layout.width,
+      height: event.nativeEvent.layout.height,
+    });
+  };
+
+  getAccessibilityLabel() {
+    const { accessibilityLabel, children } = this.props;
+    if (typeof accessibilityLabel === 'function') {
+      return accessibilityLabel(
+        this.state.currentIndex,
+        React.Children.count(children),
+      );
+    }
+    return accessibilityLabel;
+  }
+
+  handleScroll = (event: any) => {
+    const { contentOffset } = event.nativeEvent;
+    const currentIndex = this.state.width
+      ? contentOffset.x / this.state.width
+      : 0;
+    this.setState({ currentIndex });
+  };
+
+  render() {
+    const { style, children, renderPagination } = this.props;
```

Comment:
Can we put these in alphabetic order to satisfy my obsessive tendencies?

Reply:
I'll do it but I have to voice my disagreement :D. I honestly don't think this is relevant or desired, and if it is, the linter should be the one to catch/force this, otherwise it will never be a pattern throughout the code base. I also don't think it makes sense in most situations. Let's say I a have the props `width, height, style` and `alertText, errorText, successText`. I would group them together, by their "function" and not alphabetically.

Reply:
That makes sense - if you feel strongly about it I'm happy to back down!